### DISCUSSION TITLE
Time Range Search Bug

### DISCUSSION THREAD

**Discussion by cherries0524** [jstedfast/MailKit#1666] [General]
string host = "outlook.office365.com";
                int port = 993;
                List<EmailModel> result = new();

                using (var client = new ImapClient())
                {
                    await client.ConnectAsync(host, port, SecureSocketOptions.SslOnConnect);
                    await client.AuthenticateAsync(email, password);

                    var inbox = client.Inbox;
                    await inbox.OpenAsync(FolderAccess.ReadOnly);

                    // 搜索收件箱中的邮件
                    DateTime starttime = new DateTime(2023, 11, 15, 0, 0, 0, DateTimeKind.Utc);
                    DateTime endtime = new DateTime(2023, 11, 15, 23, 59, 59, DateTimeKind.Utc);
                    endtime = new DateTime(2023, 11, 16, 0, 0, 0, DateTimeKind.Utc);

                    var query = SearchQuery.DeliveredAfter(starttime).Or(SearchQuery.DeliveredOn(starttime)).And(SearchQuery.DeliveredBefore(endtime));
                    var uids = await inbox.SearchAsync(query);

                    foreach (var uid in uids)
                    {
                        var message = await inbox.GetMessageAsync(uid);
                        result.Add(new() { from = string.Join(";", message.From.Select(m => m.Name)), htmlbody = message.HtmlBody, id = uid, title = message.Subject });
                    }

                    await client.DisconnectAsync(true);
                }

                return result;

There are 3 emails in Outlook, with the email times being: 2023/11/15 08:35:32+08:00，2023/11/15 07:30:13+00:00，2023/11/15 00:51:27-08:00.
When endtime = new DateTime(2023, 11, 15, 23, 59, 59, DateTimeKind.Utc) return 0 count;
When endtime = new DateTime(2023, 11, 16, 0, 0, 0, DateTimeKind.Utc) return 3 count;
There is a bug in the international email time range search.

**Comment by jstedfast** [comment]
Hi @cherries0524 

According to the documentation for the [DeliveredAfter](http://www.mimekit.net/docs/html/M_MailKit_Search_SearchQuery_DeliveredAfter.htm), [DeliveredOn](http://www.mimekit.net/docs/html/M_MailKit_Search_SearchQuery_DeliveredOn.htm), and [DeliveredBefore](http://www.mimekit.net/docs/html/M_MailKit_Search_SearchQuery_DeliveredBefore.htm) queries:

> The resolution of this search query does not include the time.

These SearchQuery methods map to IMAP's `SINCE`, `ON`, and `BEFORE` SEARCH query operators (respectively) which are defined in [RFC3501, Section 6.4.4](https://www.rfc-editor.org/rfc/rfc3501#section-6.4.4):

```
      BEFORE <date>
         Messages whose internal date (disregarding time and timezone)
         is earlier than the specified date.

      ON <date>
         Messages whose internal date (disregarding time and timezone)
         is within the specified date.

      SINCE <date>
         Messages whose internal date (disregarding time and timezone)
         is within or later than the specified date.
```

I know it's confusing and non-intuitive due to the fact that these query methods take a `DateTime` (which would insinuate that the Time portion is taken into consideration), but unfortunately, the `System.DateOnly` type did not exist when I first wrote these methods (and still only exists in .NET >= 6.0 while MimeKit and MailKit need to support .NET Framework).

If you want to search for messages with the level of specificity that you are trying to do, you'll need an IMAP server that supports the [WITHIN](https://www.rfc-editor.org/rfc/rfc5032) extension.

You can check for this extension by doing:

```csharp
if (client.Capabilities.HasFlag(ImapCapabilities.Within))
```

If the IMAP server supports `WITHIN`, then you can use [SearchQuery.OlderThan(int seconds)](http://www.mimekit.net/docs/html/M_MailKit_Search_SearchQuery_OlderThan.htm) and/or [SearchQuery.YoungerThan(int seconds)](http://www.mimekit.net/docs/html/M_MailKit_Search_SearchQuery_YoungerThan.htm) queries.

```csharp
DateTime starttime = new DateTime(2023, 11, 15, 0, 0, 0, DateTimeKind.Utc);
DateTime endtime = new DateTime(2023, 11, 15, 23, 59, 59, DateTimeKind.Utc);
var query = SearchQuery.YoungerThan((int) (DateTime.Utc.Now - starttime).TotalSeconds).And(SearchQuery.OlderThan((int) (DateTime.Utc.Now - endtime).TotalSeconds));
```

**Comment by cherries0524** [comment]
Hi @jstedfast 
The time is variable (between sending and receiving requests), and the cycle of obtaining email results over time is not without any interval.
And Outlook does not support Within.

**Reply by jstedfast** [reply]
Then your only option is to search for a wider time range and then manually filter the results by examining the IMessageSummary.InternalDate value to see if it is within your time frame.

For example:

```csharp
DateTime starttime = new DateTime(2023, 11, 15, 0, 0, 0, DateTimeKind.Utc);
DateTime endtime = new DateTime(2023, 11, 16, 0, 0, 0, DateTimeKind.Utc);

var query = SearchQuery.DeliveredAfter(starttime).Or(SearchQuery.DeliveredOn(starttime)).And(SearchQuery.DeliveredBefore(endtime));
var uids = await inbox.SearchAsync(query);

var summaries = await inbox.FetchAsync(uids, MessageSummaryItems.UniqueId | MessageSummaryItems.InternalDate);

foreach (var summary in summaries) {
    if (summary.InternalDate >= realStartTime && summary.InternalDate <= realEndTime) {
        var message = await inbox.GetMessageAsync(summary.UniqueId);
    }
}
```

**Reply by cherries0524** [reply]
OK
